### PR TITLE
Security guidelines

### PR DESCRIPTION
**Pull request by adamnfish**
Some recommendations / guidelines relating to securing our applications. These are all based on real issues we've faced and will probably be an evolving list as we find new areas of concern and best practices move on.

### CODE REVIEW

**Comment by rrees** [comment]
:+1: HTTPS everywhere is going to require a lot infrastructure changes though

File: security.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+to get pointers on areas that might need extra attention. Revisit it
+from time to time to stay on top of changes in the application and the
+team.
+
+## AWS
+
+**Do not use permanent AWS credentials.** Applications running in AWS
+should use `assumeRole` to gain access to other AWS resources. For
+local development you should use AWS profiles to manage credentials
+and these should be temporary credentials (e.g. provided by Janus).
```
Comment (theefer) [inline security.md]: I might be wrong, but I suspect this is one of the recommendations which currently most projects do **not** follow. Are there many practical examples of this, in particular the Janus-only DEV credentials?

**Comment by adamnfish** [comment]
@rrees not sure I agree. We already do this naturally for pretty much every API and website we build, I'd be interested to hear IRL if you can think of any exceptions to this? It does need clarifying that I don't necessarily include traffic inside an AWS account (i.e. load balancer -> instances) here.

**Comment by theefer** [comment]
I do think there's room to automate SSL cert management a lot more (Let's Encrypt?) than the current manual process with cert-magic and emails to Enterprise, in order for it to be much less of a burden when dealing with lots of smaller services/projects.

File: security.md (hunk 1 above)
Comment (adamnfish) [inline security.md]: Yep, there are quite a few examples of where this isn't done at the moment. We're slowly moving existing projects away from permanent credentials (particularly in the tools team which has multiple accounts) so there are quite a few examples in there. I think Lindsey and Simon would be best placed to point you at their examples? More than the existing stuff though, this recommendation is important for teams that are thinking about how to design new applications.

**Comment by adamnfish** [comment]
@theefer yep, hopefully this will encourage us to solve those problems in a better way. It's certainly already possible because pretty much everything we do is already secure, but there's definitely some tasty fruit that's tantalisingly in reach here.

File: security.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+**Disk encryption.** All machines that you do work on must have disk
+encryption enabled.
+
+**2FA.** Whenever a service offers 2FA, you should enable it.
+
+**Use a password manager**. Don't reuse simple passwords across
+services. A password manager is a good way to help use strong
+passwords. The `pwgen` command can be used to generate good passwords,
+especially with the `-s` and `-y` flags (`pwgen -sy <length>`).
```
Comment (satterly) [inline security.md]: To install use `brew install pwgen` on a Mac.

File: security.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+## Application development
+
+**Keep secrets out of repositories.** Application secrets must not be
+stored in version control, even if the repo is "private". To prevent
+accidental leakage, files containing secrets should be stored away
+from the project, completely outside of the source tree.
+
+**Use the OWASP Security Knowledge Framework**. Before releasing a new
+application make sure you run through the Security Knowledge Framework
```
Comment (satterly) [inline security.md]: Add a link? I assume you mean these guys ... https://www.securityknowledgeframework.org/

Comment (adamnfish) [inline security.md]: I wrote this a couple of days ago, before this link existed https://skf.gutools.co.uk/ Now that it works you're absolutely right. There'll be more info to follow on this subject.

File: security.md (hunk 2 above)
Comment (adamnfish) [inline security.md]: I'm not sure "how to install software on your dev machine" belongs in the security section of the recommendations repo. I reckon people will work this out but maybe if a few mac users feel like this is necessary I can add a note to that effect?

**Comment by rrees** [comment]
@adamnfish I thought this did refer to HTTPS within accounts. I agree that it is a sensible rule for external termination.

**Comment by cb372** [comment]
Maybe add an AWS-specific note reminding people that they should make rules (e.g. IAM policies, security group rules) as restrictive as possible, only widening them later when the need arises. It's often tempting to do the opposite, leaving everything open for the sake of developer convenience, then not getting around to tightening them up later.

**Comment by philwills** [comment]
I'm going to merge this, as I'd rather have it in that not and we can improve on it later.

**Pull request merged**